File: app/backend/repeat_tool_guard.py
"""repeat_tool_guard.py — 循环死锁与重复调用主动守卫。

职责：
1. 在 AgentLoop 工具执行流中实时监控工具调用指纹；
2. 识别连续相似调用且产出无进展的死循环模式；
3. 生成建议性脱困提示（Advisory Context），引导模型跳出局部死锁；
4. 输出结构化故障指纹（Failure Signature），直接作为高质量负反馈信号喂入 GEPA 自进化引擎。
"""
from __future__ import annotations

import hashlib
import json
import time
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional


def _normalize_args_fingerprint(args: Any) -> str:
    """提取参数的规范化哈希指纹，忽略字典键序与空值微小抖动。"""
    if not isinstance(args, dict):
        return str(args)
    # 提取核心参数键值
    clean_dict = {}
    for k, v in sorted(args.items()):
        if k in ("session_id", "timestamp", "request_id"):
            continue
        if isinstance(v, str):
            clean_dict[k] = v.strip()
        elif isinstance(v, (int, float, bool)):
            clean_dict[k] = v
        elif isinstance(v, (list, dict)):
            clean_dict[k] = json.dumps(v, sort_keys=True, ensure_ascii=False)
    serialized = json.dumps(clean_dict, sort_keys=True, ensure_ascii=False)
    return hashlib.md5(serialized.encode("utf-8")).hexdigest()[:12]


@dataclass
class ToolCallEntry:
    tool_name: str
    args_fingerprint: str
    args_summary: str
    timestamp: float = field(default_factory=time.time)
    outcome_ok: bool = True
    outcome_length: int = 0


class RepeatToolGuard:
    """工具循环死锁检测器与脱困建议生成器。"""

    def __init__(self, max_window_size: int = 10, repeat_threshold: int = 3):
        self.max_window_size = max_window_size
        self.repeat_threshold = repeat_threshold
        self.history: List[ToolCallEntry] = []
        self._last_alert_fingerprint: Optional[str] = None

    def record_call(self, tool_name: str, args: dict, outcome_ok: bool = True, outcome_content: str = "") -> None:
        """记录一次工具调用的指纹。"""
        fp = _normalize_args_fingerprint(args)
        summary = str(args)[:100]
        entry = ToolCallEntry(
            tool_name=tool_name,
            args_fingerprint=fp,
            args_summary=summary,
            timestamp=time.time(),
            outcome_ok=outcome_ok,
            outcome_length=len(str(outcome_content or "")),
        )
        self.history.append(entry)
        if len(self.history) > self.max_window_size:
            self.history.pop(0)

    def check_loop(self) -> Optional[Dict[str, Any]]:
        """检查当前窗口是否命中工具死循环模式。"""
        if len(self.history) < self.repeat_threshold:
            return None

        # 检查尾部连续调用
        tail = self.history[-self.repeat_threshold:]
        target_tool = tail[0].tool_name
        target_fp = tail[0].args_fingerprint

        # 如果尾部 N 次全都是同一个工具，且参数指纹一致或相似
        is_exact_repeat = all(e.tool_name == target_tool and e.args_fingerprint == target_fp for e in tail)

        # 或者滑动窗口内同一工具、同一参数调用次数超过阈值
        count = sum(1 for e in self.history if e.tool_name == target_tool and e.args_fingerprint == target_fp)

        if (is_exact_repeat or count >= self.repeat_threshold) and target_fp != self._last_alert_fingerprint:
            self._last_alert_fingerprint = target_fp
            return {
                "detected": True,
                "tool_name": target_tool,
                "repeat_count": count,
                "args_summary": tail[-1].args_summary,
                "advisory_message": (
                    f"【系统防死循环守卫提示】: 你已连续 {count} 次执行类似的 `{target_tool}` 操作且未取得新突破。"
                    "请停止重复盲试！请尝试：1) 更换排查关键词或检索策略；2) 查看目录树或文件结构；3) 向用户澄清或说明当前困境。"
                ),
                "failure_signature": f"repeat_loop:{target_tool}:{target_fp}",
            }
        return None

    def reset(self) -> None:
        """重置守卫历史。"""
        self.history.clear()
        self._last_alert_fingerprint = None

File: src/core/execution/RepeatToolGuard.ts
/**
 * RepeatToolGuard.ts — Loop Detector with MD5 Fingerprinting & Advisory Context
 *
 * Responsibilities:
 * 1. Monitor tool execution fingerprints in real-time within the AgentLoop.
 * 2. Detect deadlock loops where the agent repeatedly runs identical or near-identical tool calls.
 * 3. Generate Chinese advisory guidance prompts to help the agent break out of local optima.
 * 4. Emit structured failure signatures (`repeat_loop:{tool_name}:{fingerprint}`) for GEPA self-evolution.
 */

import crypto from 'node:crypto'

export interface ToolCallEntry {
  toolName: string
  argsFingerprint: string
  argsSummary: string
  timestamp: number
  outcomeOk: boolean
  outcomeLength: number
}

export interface LoopAlert {
  detected: boolean
  toolName: string
  repeatCount: number
  argsSummary?: string
  advisoryMessage: string
  failureSignature: string
}

/**
 * Normalizes tool arguments and computes an MD5 hash fingerprint.
 * Strips volatile/ephemeral fields and normalizes whitespace and object key orders.
 */
export function normalizeArgsFingerprint(argsOrToolName: any, optionalArgs?: any): string {
  let toolName = ''
  let args: any = argsOrToolName

  if (typeof argsOrToolName === 'string' && optionalArgs !== undefined) {
    toolName = argsOrToolName
    args = optionalArgs
  } else if (typeof argsOrToolName === 'object' && argsOrToolName !== null && optionalArgs === undefined) {
    args = argsOrToolName
  }

  if (!args || typeof args !== 'object') {
    const raw = String(args || '')
    const hash = crypto.createHash('md5').update(raw).digest('hex').slice(0, 12)
    return toolName ? `${toolName}:${hash}` : hash
  }

  const cleanObj: Record<string, any> = {}
  const ignoredKeys = new Set([
    'session_id',
    'sessionId',
    'timestamp',
    'request_id',
    'requestId',
    'call_id',
    'callId',
    'toolSummary',
    'toolAction',
  ])

  function sanitizeValue(val: any): any {
    if (val === null || val === undefined) return val
    if (typeof val === 'string') {
      return val.trim().replace(/\s+/g, ' ')
    }
    if (typeof val === 'number' || typeof val === 'boolean') {
      return val
    }
    if (Array.isArray(val)) {
      return val.map(sanitizeValue)
    }
    if (typeof val === 'object') {
      const sortedSub: Record<string, any> = {}
      for (const k of Object.keys(val).sort()) {
        if (!ignoredKeys.has(k) && !k.startsWith('_')) {
          sortedSub[k] = sanitizeValue(val[k])
        }
      }
      return sortedSub
    }
    return String(val)
  }

  for (const k of Object.keys(args).sort()) {
    if (!ignoredKeys.has(k) && !k.startsWith('_')) {
      cleanObj[k] = sanitizeValue(args[k])
    }
  }

  const serialized = JSON.stringify(cleanObj)
  const hash = crypto.createHash('md5').update(`${toolName ? toolName + ':' : ''}${serialized}`).digest('hex').slice(0, 12)
  return toolName ? `${toolName}:${hash}` : hash
}

export class RepeatToolGuard {
  private maxWindowSize: number
  private repeatThreshold: number
  private history: ToolCallEntry[] = []
  private alertedFingerprints = new Set<string>()

  constructor(maxWindowSize = 10, repeatThreshold = 3) {
    this.maxWindowSize = maxWindowSize
    this.repeatThreshold = repeatThreshold
  }

  public recordCall(
    toolName: string,
    args: any,
    outcomeOk = true,
    outcomeContent: string | any = ''
  ): void {
    const fp = normalizeArgsFingerprint(toolName, args)
    const summary = typeof args === 'object' ? JSON.stringify(args).slice(0, 100) : String(args).slice(0, 100)

    const entry: ToolCallEntry = {
      toolName,
      argsFingerprint: fp,
      argsSummary: summary,
      timestamp: Date.now(),
      outcomeOk,
      outcomeLength: typeof outcomeContent === 'string' ? outcomeContent.length : JSON.stringify(outcomeContent || '').length,
    }

    this.history.push(entry)
    if (this.history.length > this.maxWindowSize) {
      this.history.shift()
    }
  }

  public checkLoop(): LoopAlert | null {
    if (this.history.length < this.repeatThreshold) {
      return null
    }

    // Check tail consecutive repeat
    const tail = this.history.slice(-this.repeatThreshold)
    const targetTool = tail[0].toolName
    const targetFp = tail[0].argsFingerprint

    const isExactRepeat = tail.every(
      (e) => e.toolName === targetTool && e.argsFingerprint === targetFp
    )

    // Check frequency count in sliding window
    const count = this.history.filter(
      (e) => e.toolName === targetTool && e.argsFingerprint === targetFp
    ).length

    if ((isExactRepeat || count >= this.repeatThreshold) && !this.alertedFingerprints.has(targetFp)) {
      this.alertedFingerprints.add(targetFp)
      const lastSummary = tail[tail.length - 1].argsSummary

      return {
        detected: true,
        toolName: targetTool,
        repeatCount: count,
        argsSummary: lastSummary,
        advisoryMessage: `【系统防死循环守卫提示】: 你已连续 ${count} 次执行类似的 \`${targetTool}\` 操作且未取得新突破。请停止重复盲试！请尝试：1) 更换排查关键词或检索策略；2) 查看目录树或文件结构；3) 向用户澄清或说明当前困境。`,
        failureSignature: `repeat_loop:${targetFp.startsWith(targetTool + ':') ? targetFp : `${targetTool}:${targetFp}`}`,
      }
    }

    return null
  }

  public getHistory(): ToolCallEntry[] {
    return [...this.history]
  }

  public reset(): void {
    this.history = []
    this.alertedFingerprints.clear()
  }
}
